webrender/src/device/gfx/image.rs

```diff
@@ -24,6 +26,213 @@ const DEPTH_RANGE: hal::image::SubresourceRange = hal::image::SubresourceRange {
 /// The Vulkan spec states: bufferOffset must be a multiple of 4 for VkBufferImageCopy
 /// https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VkBufferImageCopy
 const BUFFER_COPY_ALIGNMENT: i32 = 4;
+const RENDER_TARGET_MEMORY_SIZE: u64 = 128 << 20; // 128 MB
+
+type LinearMemoryId = usize;
+struct MemoryRange(std::ops::Range<hal::buffer::Offset>);
+
+impl MemoryRange {
+    fn has_size(&self, allocation_size: hal::buffer::Offset) -> bool {
+        (self.0.end - self.0.start) >= allocation_size
+    }
+
+    fn split(self, size: hal::buffer::Offset) -> (Self, Self) {
+        (
+            MemoryRange(self.0.start .. self.0.start + size),
+            MemoryRange(self.0.start + size .. self.0.end)
+        )
+    }
+
+    fn can_merge(&self, other: &Self) -> bool {
+        self.0.end == other.0.start
+        || self.0.start == other.0.end
+    }
+
+    fn merge(self, other: Self) -> Self {
+        MemoryRange(
+            self.0.start.min(other.0.start)
+                .. self.0.end.max(other.0.end)
+        )
+    }
+}
+
+struct LinearMemoryBlock<B: hal::Backend> {
+    memory_block: MemoryBlock<B>,
+    alignment: u64,
+    type_mask: u64,
+    free_chunks: Vec<MemoryRange>,
```

Comment:
My understanding is that the whole point of a linear allocator ("Linear" in the `LinearMemoryBlock`) that we don't need to track any free chunks, otherwise it's becoming a general-purpose allocator.

Reply:
That's true, and that was the case before I introduced the tracking of free chunks(which is required to reuse memory between passes), I just forget to change the name of the allocator and the block. I will change it to avoid further confusion.